src/back-end/lib/db.ts

```diff
@@ -654,3 +656,434 @@ export async function hasFilePermission(connection: Connection, session: Session
+      .transacting(trx)
+      .insert({
+        id: generateUuid(),
+        opportunity: rootOppRecord.id,
+        createdAt: now,
+        createdBy: session.user?.id,
+        status: CWUOpportunityStatus.Draft,
+        note: ''
+      }, '*');
+
+    // Create attachment records
+    attachments?.forEach(async attachment => {
+      await connection('cwuOpportunityAttachments')
+        .transacting(trx)
+        .insert({
+          opportunityVersion: oppVersionRecord.id,
+          file: attachment
+        }, '*');
+    });
+
+    const dbResult = await readOneCWUOpportunity(trx, rootOppRecord.id, session);
+    if (isInvalid(dbResult) || !dbResult.value) {
+      throw new Error('unable to create opportunity');
+    }
+    return dbResult.value;
+  });
+
+  return valid(result);
+});
+
+export async function isCWUOpportunityAuthor(connection: Connection, user: User, id: Id): Promise<boolean> {
+  try {
+    const result = await connection<RawCWUOpportunity>('cwuOpportunities')
+      .select('*')
+      .where({ id, createdBy: user.id });
+    return !!result && result.length > 0;
+  } catch (exception) {
+    return false;
+  }
+}
+
+type UpdateCWUOpportunityParams = Partial<CWUOpportunity>;
+
+export const updateCWUOpportunityVersion = tryDb<[UpdateCWUOpportunityParams, Session], CWUOpportunity>(async (connection, opportunity, session) => {
+  const now = new Date();
+  const { attachments, ...restOfOpportunity } = opportunity;
+  return valid(await connection.transaction(async trx => {
+    const [oppVersion] = await connection<OpportunityVersionRecord>('cwuOpportunityVersions')
+      .transacting(trx)
+      .insert({
+        ...restOfOpportunity,
+        opportunity: restOfOpportunity.id,
+        id: generateUuid(),
+        createdAt: now,
+        createdBy: session.user?.id
+      }, '*');
+
+    if (!oppVersion) {
+      throw new Error('unable to update opportunity');
+    }
+
+    attachments?.forEach(async attachment => {
+      await connection('cwuOpportunityAttachments')
+        .transacting(trx)
+        .insert({
+          opportunity: oppVersion.id,
+          file: attachment.id
+        });
+    });
+
+    const dbResult = await readOneCWUOpportunity(connection, oppVersion.opportunity, session);
+    if (isInvalid(dbResult) || !dbResult.value) {
+      throw new Error('unable to update opportunity');
+    }
+    return dbResult.value;
+  }));
+});
+
+interface CWUOpportunityStatusRecord {
+  id: Id;
+  opportunity: Id;
+  createdAt: Date;
+  createdBy: Id;
+  status: CWUOpportunityStatus;
+  note: string;
+}
+
+export const updateCWUOpportunityStatus = tryDb<[Id, CWUOpportunityStatus, string, Session], CWUOpportunity>(async (connection, id, status, note, session) => {
+  const now = new Date();
+  const [result] = await connection<CWUOpportunityStatusRecord>('cwuOpportunityStatuses')
+    .insert({
+      id: generateUuid(),
+      opportunity: id,
+      createdAt: now,
+      createdBy: session.user?.id,
```

Comment:
This entire operation should fail if the user is not signed in — shouldn't use optional chaining here.